docs/UA-service-API.md

```diff
@@ -0,0 +1,456 @@
+Service REST API.
+
+# Overview
+
+The User Agent Service provides two sub-components intended to back an
+abstract conception of browsing: an HTTP REST API to allow bulk access
+to data, and a light WebSocket API to push incremental changes to
+clients.  The WebSocket API will be documented separately.
+
+The abstract conception of browsing introduces two high-level
+concepts, _scopes_ and _sessions_.  Roughly speaking, a _scope_
+identifies a logical or conceptual grouping of browsing behaviour.
+For Tofino, a scope corresponds to an open browser window; for a
+mobile search activity, a scope might correspond to a single
+interaction with the main activity.  A _session_ identifies a single
+browsing context.  It takes place within a scope (put differently,
+scopes encapsulate sessions) and may have a logical _ancestor
+session_.  For Tofino, a session corresponds to a single tab; a tab
+spawned by clicking a link with "target=_blank" would correspond to a
+new session with the original tab as it's ancestor session.
+
+Scopes are currently allocated as new sessions in Tofino.
+Alternatively, use any fixed integer scope (such as scope 0).
+
+Full-text indexing is currently opt-in: see
+[POST /v1/pages/:url](#post-v1pagesurl).  The data store will return
+snippets, which for now are HTML escaped string with bold tags
+wrapping highlighted words.  It's safe to use `innerContent` and
+friends to render snippets in privileged UI.
+
+Throughout, timestamps are integer milliseconds since the Unix epoch.
+
+## URL Structure
+
+All requests will be to URLs for the form:
+
+    http[s]://<server-url>/v1/<api-endpoint>
+
+Note that:
+
+* The URL embeds a version identifier "v1"; future revisions of this
+  API may introduce new version numbers.
+* The base URL of the server may be configured on a per-client basis.
+* At this time, there is no requirement for API access to be over a
+  properly-validated HTTPS connection.
+
+## Request Format
+
+All POST requests must have a content-type of `application/json` with
+a utf8-encoded JSON body, and should specify the content-length
+header.  Most requests include body parameters, but some have query
+parameters.  Over time we intend to migrate to all body parameters, to
+be consistent with other Mozilla services.
+
+## Response Format
+
+All successful requests will produce a response with HTTP status code
+of "200" and content-type of "application/json".  The structure of the
+response body will depend on the endpoint in question.
+
+Failures due to invalid behavior from the client will produce a
+response with HTTP status code in the "4XX" range and content-type of
+"application/json".  Failures due to an unexpected situation on the
+server will produce a response with HTTP status code in the "5XX"
+range and content-type of "application/json".
+
+---
+
+TODO: diffs WebSocket API documentation
+
+---
+
+# API Endpoints
+
+* Sessions
+    * [POST /v1/session/start](#post-v1sessionstart)
```

Comment:
Are there supposed to be elements on this page with id="post-v1sessionstart" for this anchor to work?